lib/src/test/java/io/ably/lib/test/realtime/RealtimeChannelTest.java

```diff
@@ -196,6 +191,58 @@ public void attach_with_channel_params_set_options() {
         }
     }
 
+    @Test
+    public void attached_state_is_correctly_achieved() {
+        String channelName = "test_" + testParams.name;
+        AblyRealtime ably = null;
+        AblyRest ablyRest;
+
+        try {
+            ClientOptions opts = createOptions(testVars.keys[0].keyStr);
+            ably = new AblyRealtime(opts);
+            ablyRest = new AblyRest(opts);
+
+            /* wait until connected */
+            (new ConnectionWaiter(ably.connection)).waitFor(ConnectionState.connected);
+            assertEquals("Verify connected state reached", ably.connection.state, ConnectionState.connected);
+
+            final Map<String, String> params = new HashMap<>();
+            params.put("rewind", "1");
+            final ChannelOptions options = new ChannelOptions();
+            options.params = params;
+
+            Channel channel = ably.channels.get(channelName, options);
+            ablyRest.channels.get(channelName).publish("", "test-message");
+
+            // channel is in initialized state when created
+            assertEquals(ChannelState.initialized, channel.state);
+
+            Helpers.MessageWaiter messageWaiter = new Helpers.MessageWaiter(channel);
+            // channel is not attached yet
+            assertEquals(ChannelState.attaching, channel.state);
+
+            messageWaiter.waitFor(1);
+            // when message is received channel state is already in attached state
+            assertEquals(ChannelState.attached, channel.state);
+
+            ably.channels.release(channelName);
+
+            // verify there are no channels after realising
+            assertFalse(ably.channels.entrySet().iterator().hasNext());
+            Thread.sleep(5000);
```

Comment:
Why is this `sleep()` needed?

Edit: I've read the issue discussion now that this was based on. Have you managed to get this test to fail before the fix?

Reply:
No, added sleeping didn't changed anything but I left it here for demonstration.